# Week 3 Exercises

# Basic 1

In [13]:
def removeDuplicates(inputList):
    result = []
    # loop through the input list
    for i in inputList:
        foundMatch = False
        # check if the element is in the results
        for j in result:
            if i == j:
                foundMatch = True
        # if we can't find the element in the reuslts, add it
        if not foundMatch:
            result += [i]
    return result

print(removeDuplicates([0,1,2,2,3,4,5,8,10,10]))
print(removeDuplicates(['apple','apple','orange','banana']))
print(removeDuplicates([]))


[0, 1, 2, 3, 4, 5, 8, 10]
['apple', 'orange', 'banana']
[]


# Basic 2

In [26]:
def isNumberWithinList(orderedInputList, number):
    # assuming the input is in order, the number must be
    # greater than the first entry in the list and smaller than
    # the last
    if (number > orderedInputList[0]) and (number < orderedInputList[len(orderedInputList)-1]):
        return True
    return False

print(isNumberWithinList([10,20,30], 15))
print(isNumberWithinList([1,2,3], 5))
print(isNumberWithinList([1,2], 1.5))
print(isNumberWithinList([1], 1.5))
print(isNumberWithinList([5,10], 8))


True
False
True
False
True


# Basic 3

In [4]:
a = [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
b = [i for i in a if i%2 == 0]
print (b)

[4, 16, 36, 64, 100]


# Advanced 1

In [18]:
import random

def guessMyNumberGame():
    closeGuess = 3
    wayOffGuess = 100
    myNum = random.randint(0, 200)
    # for comparison question:
    # myNum = 138
    while True:
        userGuess = input('Guess my number: ')
        delta = myNum - int(userGuess)
        if delta == 0:
            print ('bingo!')
            return
        if delta > wayOffGuess:
            print ('way higher')
        elif delta > closeGuess:
            print ('higher')
        elif delta < -wayOffGuess:
            print ('way lower')
        elif delta < -closeGuess:
            print ('lower')
        else:
            print ('close')

guessMyNumberGame()

Guess my number: 100
higher
Guess my number: 120
higher
Guess my number: 140
close
Guess my number: 137
close
Guess my number: 138
bingo!


# Advanced 2

In [25]:
import random


def guessYourNumberGame():
    # when the user says it's close, this means within this much
    closeGuess = 3
    # the lowest possible User number
    lowestUser = 0
    # the highest possible User number
    highestUser = 200
    # list of possible numbers
    possibleList = list(range(lowestUser, highestUser + 1))
    # the initial program Guess, start in the middle
    programGuess = possibleList[len(possibleList) // 2]
    # max guesses
    maxGuesses = 10
    curGuess = 0

    print('Think of an integer from ' + str(lowestUser) + ' to ' + str(highestUser), end='')
    print(' and I will try to guess it!')
    print('Reply with one of the following:')
    print('1 = way higher, 2 = higher, 3 = close (within ' + str(closeGuess) + ')')
    print('4 = way lower, 5 = lower, 0 = bingo!')

    while curGuess < maxGuesses:
        # counter for the number of guesses
        curGuess += 1
        # get reply from user
        userResponse = int(input('Is your number ' + str(programGuess) + '? '))
        if userResponse == 0:
            print('Bingo!')
            return

        # if the user says higher, then we can discard all lower possibilities
        if userResponse == 1 or userResponse == 2:
            possibleList = [i for i in possibleList if i > programGuess]
        # if the user says lower, then we can discard all higher possibilities
        if userResponse == 4 or userResponse == 5:
            possibleList = [i for i in possibleList if i < programGuess]
        # if the user says close, discard anything further than close to the guess
        if userResponse == 3:
            possibleList = [i for i in possibleList
                            if (programGuess + closeGuess) >= i >= (programGuess - closeGuess) and i != programGuess]

        # check if there's only 1 element in the possibility list, if so it must be it
        if len(possibleList) == 1:
            print('Your number must be ' + str(possibleList[0]) + '!')
            return
        # check for an empty list, something must have gone wrong!
        if len(possibleList) == 0:
            print('I can\'t figure out your number!')
            return

        # now create the guess from the shortened list
        if userResponse == 1 or userResponse == 4:
            # guess the middle of the list of they say way higher or way lower
            programGuess = possibleList[len(possibleList) // 2]
        elif userResponse == 2:
            # guess 1/4 of the way into the list if they say higher
            programGuess = possibleList[len(possibleList) // 4]
        elif userResponse == 5:
            # guess 3/4 of the way into the list if they say lower
            programGuess = possibleList[3 * len(possibleList) // 4]
        elif userResponse == 3:
            # close, so take a random entry from the possible list because we just know it's close
            randomIndex = random.randint(0, len(possibleList) - 1)
            programGuess = possibleList[randomIndex]
        else:
            print('unknown option!')

    print('Too many guesses!')


guessYourNumberGame()


Think of an integer from 0 to 200 and I will try to guess it!
Reply with one of the following:
1 = way higher, 2 = higher, 3 = close (within 3)
4 = way lower, 5 = lower, 0 = bingo!
Is your number 100? 1
Is your number 151? 2
Is your number 164? 3
Is your number 167? 5
Is your number 165? 3
Is your number 163? 3
Is your number 162? 0
Bingo!


# Advanced 3

In [20]:
import random

def guessMyNumberGameExtra():
    maxGuessNum = 10
    currentGuessNum = 0
    closeGuess = 3
    wayOffGuess = 100
    myNum = random.randint(0, 200)
    # for comparison question:
    # myNum = 138
    while currentGuessNum < maxGuessNum:
        userGuess = input('Guess my number: ')
        delta = myNum - int(userGuess)
        currentGuessNum += 1
        wayOff = False
        if delta == 0:
            print ('bingo!')
            return
        if delta > wayOffGuess:
            wayOff = True
            print ('way higher')
        elif delta > closeGuess:
            print ('higher')
        elif delta < -wayOffGuess:
            wayOff = True
            print ('way lower')
        elif delta < -closeGuess:
            print ('lower')
        else:
            print ('close')
        if not wayOff and currentGuessNum >= 2:
            if myNum%2 == 0:
                print ('Hint: the number is even!')
            else:
                print ('Hint: the number is odd!')
            if myNum%3 == 0:
                print ('Hint: the number is divisible by 3!')
            else:
                print ('Hint: the number is not divisible by 3')
    print('too many guesses!')
                

guessMyNumberGameExtra()

Guess my number: 100
higher
Guess my number: 120
higher
Hint: the number is even!
Hint: the number is divisible by 3!
Guess my number: 138
bingo!


When you give a hint of even and divisible by three, it reduces the number of guesses when you get close to the number since it reduces the possibilities.  In this case, if the number is 138 and you follow a +20 guess strategy, the number of guesses is only 3 since the nearest even and divisible by 3 value that is +20 from 120 is 138.  This is a lot less than the number of guesses in Advanced 1 which was 5.

# Advanced 4

In [27]:
import random


def guessYourNumberGameExtra():
    # when the user says it's close, this means within this much
    closeGuess = 3
    # the lowest possible User number
    lowestUser = 0
    # the highest possible User number
    highestUser = 200
    # list of possible numbers
    possibleList = list(range(lowestUser, highestUser + 1))
    # the initial program Guess, start in the middle
    programGuess = possibleList[len(possibleList) // 2]
    # max guesses
    maxGuesses = 10
    curGuess = 0
    hintAfterGuess = 2
    # hints
    askedEven = False
    askedDivisibleByThree = False

    print('Think of an integer from ' + str(lowestUser) + ' to ' + str(highestUser), end='')
    print(' and I will try to guess it!')
    print('Reply with one of the following:')
    print('1 = way higher, 2 = higher, 3 = close (within ' + str(closeGuess) + ')')
    print('4 = way lower, 5 = lower, 0 = bingo!')

    while curGuess < maxGuesses:
        # counter for the number of guesses
        curGuess += 1
        # get reply from user
        userResponse = int(input('Is your number ' + str(programGuess) + '? '))
        if userResponse == 0:
            print('Bingo!')
            return

        # ask the odd even hint and remove impossible elements from list
        if curGuess >= hintAfterGuess and not askedEven:
            userResponseStr = input('Is your number even? (y/n): ')
            if userResponseStr.lower() == 'y':
                possibleList = [i for i in possibleList if i % 2 == 0]
            else:
                possibleList = [i for i in possibleList if i % 2 != 0]
            askedEven = True
        elif curGuess >= hintAfterGuess and not askedDivisibleByThree:
            # ask the divisible by 3 hint and remove impossible elements from list
            userResponseStr = input('Is your number divisible by 3? (y/n): ')
            if userResponseStr.lower() == 'y':
                possibleList = [i for i in possibleList if i % 3 == 0]
            else:
                possibleList = [i for i in possibleList if i % 3 != 0]
            askedDivisibleByThree = True

        # if the user says higher, then we can discard all lower possibilities
        if userResponse == 1 or userResponse == 2:
            possibleList = [i for i in possibleList if i > programGuess]
        # if the user says lower, then we can discard all higher possibilities
        if userResponse == 4 or userResponse == 5:
            possibleList = [i for i in possibleList if i < programGuess]
        # if the user says close, discard anything further than close to the guess
        if userResponse == 3:
            possibleList = [i for i in possibleList
                            if (programGuess + closeGuess) >= i >= (programGuess - closeGuess)
                            and i != programGuess]

        # check if there's only 1 element in the possibility list, if so it must be it
        if len(possibleList) == 1:
            print('Your number must be ' + str(possibleList[0]) + '!')
            return
        # check for an empty list, something must have gone wrong!
        if len(possibleList) == 0:
            print('I can\'t figure out your number!')
            return

        # now create the guess from the shortened list
        if userResponse == 1 or userResponse == 4:
            # guess the middle of the list of they say way higher or way lower
            programGuess = possibleList[len(possibleList) // 2]
        elif userResponse == 2:
            # guess 1/4 of the way into the list if they say higher
            programGuess = possibleList[len(possibleList) // 4]
        elif userResponse == 5:
            # guess 3/4 of the way into the list if they say lower
            programGuess = possibleList[3 * len(possibleList) // 4]
        elif userResponse == 3:
            # close, so take a random entry from the possible list because we just know it's close
            randomIndex = random.randint(0, len(possibleList) - 1)
            programGuess = possibleList[randomIndex]
        else:
            print('unknown option!')

    print('Too many guesses!')


guessYourNumberGameExtra()


Think of an integer from 0 to 200 and I will try to guess it!
Reply with one of the following:
1 = way higher, 2 = higher, 3 = close (within 3)
4 = way lower, 5 = lower, 0 = bingo!
Is your number 100? 1
Is your number 151? 2
Is your number even? (y/n): y
Is your number 164? 3
Is your number divisible by 3? (y/n): y
Your number must be 162!


When comparing the total number of questions to the same game without the hints for number 162, with hints we have a total of 5 questions but only 3 guesses, while without the hints we had 7 guesses (Advanced 2).  But since both games use a random guess when the answer is close, this difference could vary.